app_check/src/desktop/app_check_desktop.cc

```diff
@@ -47,8 +67,33 @@ void AppCheckInternal::SetTokenAutoRefreshEnabled(
 
 Future<AppCheckToken> AppCheckInternal::GetAppCheckToken(bool force_refresh) {
   auto handle = future()->SafeAlloc<AppCheckToken>(kAppCheckFnGetAppCheckToken);
-  AppCheckToken token;
-  future()->CompleteWithResult(handle, 0, token);
+  if (!force_refresh && HasValidCacheToken()) {
+    // If the cached token is valid, and not told to refresh, return the cache
+    future()->CompleteWithResult(handle, 0, cached_token_);
+  } else {
+    // Get a new token, and pass the result into the future.
+    AppCheckProvider* provider = nullptr;
+    if (g_provider_factory && app_) {
+      provider = g_provider_factory->CreateProvider(app_);
```

Comment:
It seems a little bit wrong to me that we call CreateProvider for every GetToken request.

For custom providers, it is possible that the factory does not cache the provider internally, and so we might be making a new provider every time.

Since we already have a separate instance of AppCheckInternal for each app, and we assume that SetFactory is called before we create an instance of AppCheck, I think it might make sense to have provider as a private variable of AppCheckInternal and call CreateProvider in the constructor of AppCheckInternal.

The only potential downside of that that I see is if we want to allow changing factory after initializing app-check, that would become more complicated.

Reply:
That makes sense, I was definitely waffling a bit on whether to just keep the provider or not.  Will change.

Reply:
Sounds good.
Please update this line to use the newly created GetProvider method.